src/Core/Migration/V6_7/Migration1728040169AddPrimaryOrderDelivery.php

```diff
@@ -0,0 +1,70 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\Migration\V6_7;
+
+use Doctrine\DBAL\ArrayParameterType;
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\ParameterType;
+use Shopware\Core\Framework\Log\Package;
+use Shopware\Core\Framework\Migration\MigrationStep;
+
+/**
+ * @internal
+ */
+#[Package('checkout')]
+class Migration1728040169AddPrimaryOrderDelivery extends MigrationStep
+{
+    public function getCreationTimestamp(): int
+    {
+        return 1728040169;
+    }
+
+    public function update(Connection $connection): void
+    {
+        // No foreign key set both from order -> (primary) order delivery on
+        // purpose so the DAL can handle the circular reference. We have a similar situation with the order and order
+        // address.
+        if (!$this->columnExists($connection, 'order', 'primary_order_delivery_id')) {
+            $connection->executeStatement(
+                'ALTER TABLE `order`
+                ADD COLUMN `primary_order_delivery_id` BINARY(16) NULL DEFAULT NULL,
+                ADD COLUMN `primary_order_delivery_version_id` BINARY(16) NULL DEFAULT NULL,
+                ADD UNIQUE INDEX `uidx.order.primary_order_delivery` (`id`, `version_id`, `primary_order_delivery_id`);'
+            );
+
+            $batchSize = 1000;
+
+            do {
+                $ids = $connection->fetchFirstColumn(
```

Comment:
I don't think you need to fetch the IDs first. have a look at https://github.com/shopware/shopware/blob/trunk/src/Core/Migration/V6_4/Migration1648709176CartCompression.php

just have a look at the affected rows and do the update until there is no `primary_order_delivery_id` with `null` left

Reply:
We tried this. But the issue is, that we have an INNER JOIN here and MySQL does not allow Limits on top level UPDATE statements with joins. You get sth like this: `An exception occurred while executing a query: SQLSTATE[HY000]: General error: 1221 Incorrect usage of UPDATE and LIMIT`

We could rebuild this in a subquery to work with a RetryableQuery  I guess, but this makes the query even more complex and more unmaintainable, which is why we opted to split this into two calls, fetching the IDs first.

Reply:
ah okay, thanks for explaining :bulb: :+1: